garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+        to `__init__`, will construct this model.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        return self._inputs[0]
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        return self._outputs[0]
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        return self._inputs
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        return self._outputs
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        raise NotImplementedError
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        args, kwargs = self.init_spec
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
+
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
+        out._default_parameters = data['parameters']
+        self.__dict__.update(out.__dict__)
+
+
+class TfModel(AutoPickableModel):
+    r"""
+    Model class for TensorFlow.
+
+    This class is a parent-child design.
+    The parent model contains the parameters, i.e. the structure, and
+    two child model are constructed with input_1 and input_2. Inputs and
+    outputs of the parent model will be binded with the default child, which
+    is the first child constructed.
+
+    When new child model is created, it reuses the parameter from the
+    parent model and can be accessed by calling model.child_2, where
+    child_2 is the name given when building this particular child model.
+    If a child is constructed without given a name, the name "default" will
+    be used.
+
+    The design is illustrated as the following:
+
+         input_1                      input_2
+            |                            |
+    ============== Parent (TfModel)===============
+    |       |                            |       |
+    |       |          Parameters        |       |
+    |    ===========  /           \  =========== |
+    |    | Child 1 | /             \ | Child 2 | |
+    |    |(TfModel)|/               \|(TfModel)| |
+    |    ===========                 =========== |
+    |       |                            |       |
+    ==============================================
+            |                            |
+       (model.outputs) OR                |
+    (model.default.outputs)        model.child_2.outputs
+
+    The parameters inside a model will be initialized when calling build().
+    Do not call tf.global_variable_initializers() after building a model as it
```

Comment:
I think this should be in its own paragraph and have some way of calling out it's importance